channelserver/src/meta.rs

```diff
@@ -0,0 +1,353 @@
+        sender
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use super::*;
+    use std::collections::BTreeMap;
+    
+    use http;
+
+    #[test]
+    fn test_preferred_language() {
+        let langs = preferred_languages("en-US,es;q=0.1,en;q=0.5,*;q=0.2".to_owned());
+        assert_eq!(
+            vec![
+                "en-us".to_owned(),
+                "en".to_owned(),
+                "*".to_owned(),
+                "es".to_owned(),
+                "en".to_owned(),
+            ],
+            langs
+        );
+    }
+
+    #[test]
+    fn test_get_preferred_language_element() {
+        let langs = vec![
+            "en-us".to_owned(),
+            "en".to_owned(),
+            "es".to_owned(),
+            "en".to_owned(),
+        ];
+        // Don't include the default "en" so we can test no matching languages.
+        let bad_lang = vec!["fu".to_owned()];
+        // Include the "*" so we can return any language.
+        let any_lang = vec!["fu".to_owned(), "*".to_owned(), "en".to_owned()];
+        let mut elements = BTreeMap::new();
+        elements.insert("de".to_owned(), "Kalifornien".to_owned());
+        elements.insert("en".to_owned(), "California".to_owned());
+        elements.insert("fr".to_owned(), "Californie".to_owned());
+        elements.insert("ja".to_owned(), "カリフォルニア州".to_owned());
+        assert_eq!(
+            Some("California".to_owned()),
+            get_preferred_language_element(&langs, elements.clone())
+        );
+        assert_eq!(
+            None,
+            get_preferred_language_element(&bad_lang, elements.clone())
+        );
+        // Return Dutch, since it's the first key listed.
+        assert!(get_preferred_language_element(&any_lang, elements.clone()).is_some());
+        let goof_lang = vec!["🙄💩".to_owned()];
+        assert_eq!(
+            None,
+            get_preferred_language_element(&goof_lang, elements.clone())
+        );
+    }
+
+    #[test]
+    fn test_ua() {
+        let good_header = "Mozilla/5.0 Foo";
+        let blank_header = "";
+        let mut good_headers = http::HeaderMap::new();
+        good_headers.insert(http::header::USER_AGENT, http::header::HeaderValue::from_static(good_header));
+        assert_eq!(Some(good_header.to_owned()), get_ua(&good_headers, None));
+        let mut blank_headers = http::HeaderMap::new();
+        blank_headers.insert(http::header::USER_AGENT, http::header::HeaderValue::from_static(blank_header));
+        assert_eq!(None, get_ua(&blank_headers, None));
+        let empty_headers = http::HeaderMap::new();
+        assert_eq!(None, get_ua(&empty_headers, None));
+    }
+
+    #[ignore]
+    #[test]
+    fn test_location() {
+        let test_ip = "63.245.208.195"; // Mozilla
+
+        let langs = vec!["en".to_owned()];
+        let mut sender = SenderData::default();
+        sender.remote = Some(test_ip.to_owned());
+        // TODO: either mock maxminddb::Reader or pass it in as a wrapped impl
```

Comment:
A slightly different approach is to add in a maxmind fetch to the docker image (which I am going to do anyway) and update the `test_chan` package to test the origin IP address via a spoofed header. 

Yeah, I know...